File: java/src/main/java/encapsulation/Etudiant.java
package encapsulation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Etudiant extends Personne {
    private final String numeroEtudiant;
    private double moyenne;
    private final List<Cours> listeCours;

    public Etudiant(String nom, int age, String numeroEtudiant, double moyenne) {
        super(nom, age);
        this.numeroEtudiant = numeroEtudiant;
        this.listeCours = new ArrayList<>();
        setMoyenne(moyenne);
    }

    public String getNumeroEtudiant() {
        return numeroEtudiant;
    }

    public double getMoyenne() {
        return moyenne;
    }

    public void setMoyenne(double moyenne) {
        if (moyenne < 0 || moyenne > 20) {
            throw new IllegalArgumentException("La moyenne doit être comprise entre 0 et 20.");
        }
        this.moyenne = moyenne;
    }

    public List<Cours> getListeCours() {
        return Collections.unmodifiableList(listeCours);
    }

    public void ajouterCours(Cours cours) {
        listeCours.add(cours);
    }

    @Override
    public String toString() {
        StringBuilder coursStr = new StringBuilder();
        if (listeCours.isEmpty()) {
            coursStr.append("Aucun");
        } else {
            for (int i = 0; i < listeCours.size(); i++) {
                coursStr.append(listeCours.get(i).getNomCours());
                if (i < listeCours.size() - 1) {
                    coursStr.append(", ");
                }
            }
        }

        return super.toString()
                + " | ID: " + numeroEtudiant
                + " | Moyenne: " + moyenne
                + " | Cours: " + coursStr;
    }
}

File: python/Encapsulation/src/etudiant.py
from .personne import Personne
from .cours import Cours
from typing import List

class Etudiant(Personne):
    def __init__(self, nom: str, age: int, numero_etudiant: str, moyenne: float):
        super().__init__(nom, age)
        self.numero_etudiant = numero_etudiant
        self.moyenne = moyenne
        self.liste_cours: List[Cours] = []

    @property
    def moyenne(self):
        return self._moyenne

    @moyenne.setter
    def moyenne(self, value):
        if not (0 <= value <= 20):
            raise ValueError("La moyenne doit être comprise entre 0 et 20.")
        self._moyenne = value

    def ajouter_cours(self, cours: Cours):
        self.liste_cours.append(cours)

    def __str__(self):
        cours_str = ", ".join(c.nom_cours for c in self.liste_cours) if self.liste_cours else "Aucun"
        return f"{super().__str__()} | ID: {self.numero_etudiant} | Moyenne: {self._moyenne} | Cours: {cours_str}"